pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+    res = linop_block_diagonal.to_dense()
+
+    if int_dtype is None:
+        return res
+
+    return tf.cast(res, int_dtype)
+
+
+ar.register_function("tensorflow", "block_diag", _block_diag_tf)
+
+
+def _scatter_element_add_tf(tensor, index, value):
+    import tensorflow as tf
+
+    indices = tf.expand_dims(index, 0)
+    value = tf.cast(tf.expand_dims(value, 0), tensor.dtype)
+    return tf.tensor_scatter_nd_add(tensor, indices, value)
+
+
+ar.register_function("tensorflow", "scatter_element_add", _scatter_element_add_tf)
+
+
+# -------------------------------- Torch --------------------------------- #
+
+
+ar.register_function("torch", "asarray", lambda x: _i("torch").as_tensor(x))
+ar.register_function("torch", "diag", lambda x, k=0: _i("torch").diag(x, diagonal=k))
+ar.register_function("torch", "expand_dims", lambda x, axis: _i("torch").unsqueeze(x, dim=axis))
+ar.register_function("torch", "shape", lambda x: tuple(x.shape))
+ar.register_function("torch", "gather", lambda x, indices: x[indices])
+
+ar.register_function(
+    "torch",
+    "sqrt",
+    lambda x: _i("torch").sqrt(
+        x.to(_i("torch").float64) if x.dtype in (_i("torch").int64, _i("torch").int32) else x
+    ),
+)
+
+
+def _take_torch(tensor, indices, axis=None):
+    torch = _i("torch")
+
+    if not isinstance(indices, torch.Tensor):
+        indices = torch.as_tensor(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    if indices.ndim == 1:
+        if (indices < 0).any():
+            # index_select doesn't allow negative indices
+            dim_length = tensor.size()[0] if axis is None else tensor.shape[axis]
+            indices = torch.where(indices >= 0, indices, indices + dim_length)
+
+        return torch.index_select(tensor, dim=axis, index=indices)
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[fancy_indices]
+
+
+ar.register_function("torch", "take", _take_torch)
+
+
+def _coerce_types_torch(tensors):
+    torch = _i("torch")
+    tensors = [torch.as_tensor(t) for t in tensors]
+    dtypes = {i.dtype for i in tensors}
+
+    if len(dtypes) == 1:
+        return tensors
+
+    complex_priority = [torch.complex64, torch.complex128]
+    float_priority = [torch.float16, torch.float32, torch.float64]
+    int_priority = [torch.int8, torch.int16, torch.int32, torch.int64]
+
+    complex_type = [i for i in complex_priority if i in dtypes]
+    float_type = [i for i in float_priority if i in dtypes]
+    int_type = [i for i in int_priority if i in dtypes]
+
+    cast_type = complex_type or float_type or int_type
+    cast_type = list(cast_type)[-1]
+
+    return [t.to(cast_type) for t in tensors]
+
+
+ar.register_function("torch", "coerce", _coerce_types_torch)
+
+
+def _block_diag_torch(tensors):
+    torch = _i("torch")
+    sizes = np.array([t.shape for t in tensors])
+    res = torch.zeros(np.sum(sizes, axis=0).tolist(), dtype=tensors[0].dtype)

```

Comment:
(Minor): it may be good for readability to assign the result of `np.sum(sizes, axis=0).tolist()` to an object in the previous line and use that name here in line 284.

Reply:
```suggestion
    shape = np.sum(sizes, axis=0).tolist()
    res = torch.zeros(shape, dtype=tensors[0].dtype)
```